skbio/alignment/_tabular_msa.py

```diff
@@ -716,8 +730,119 @@ def reindex(self, key=None, keys=None):
+            If no key is provided, but the MSA has keys, and no cached key
+            is found.
+
+        See Also
+        --------
+        reindex
+
+        Notes
+        -----
+        The MSA is not automatically re-aligned when a sequence is appended.
+        Therefore, this operation is not necessarily meaningful on its own.
+
+        Examples
+        --------
+        >>> from skbio import DNA, TabularMSA
+        >>> msa = TabularMSA([DNA('')])
+        >>> msa.append(DNA(''))
+        >>> msa == TabularMSA([DNA(''), DNA('')])
+        True
+
+        >>> msa = TabularMSA([DNA('', metadata={'id': 'a'})], minter='id')
+        >>> msa.append(DNA('', metadata={'id': 'b'}), minter='id')
+        >>> msa == TabularMSA([DNA('', metadata={'id': 'a'}),
+        ...                    DNA('', metadata={'id': 'b'})], minter='id')
+        True
+        """
+
+        if key is not None and minter is not None:
+            raise OperationError("Cannot provide key and minter.")
+
+        if self.has_keys():
+            if key is None and minter is None:
+                raise OperationError(
+                    "MSA has keys but no key or minter was privided.")
+            elif key is not None:
+                new_key = key
+            elif minter is not None:
+                new_key = resolve_key(sequence, minter)
+            else:
+                raise Exception("Explicitly handle impossible branch")
+            self._add_key(new_key)
+        else:
+            if key is not None:
+                raise OperationError(
+                    "key was provided but MSA does not have keys.")
+            elif minter is not None:
+                raise OperationError(
+                    "minter was provided but MSA does not have keys.")
+
+        self._add_sequence(sequence)
+
+    def _add_sequence(self, sequence):
+        msa_is_empty = not len(self)
+        dtype = type(sequence)
+        if msa_is_empty:
+            if not issubclass(dtype, IUPACSequence):
+                raise TypeError(
+                    "`sequence` must be a scikit-bio sequence object "
+                    "that has an alphabet, not type %r" % dtype.__name__)
+            self._dtype = dtype
+            self._seqs = [sequence]
+        elif dtype is not self.dtype:
+            raise TypeError(
+                "`sequence` Must match the type of any other sequences "
+                "already in the MSA. Type %r does not match type %r" %
+                (dtype.__name__, self.dtype.__name__))
+        elif len(sequence) != self.shape.position:
+            raise ValueError(
+                "`sequence` length must match the length of any other "
+                "sequences already in the MSA: %r != %r"
+                % (len(sequence), self.shape.position))
+        else:
+            self._seqs.append(sequence)
+
+    def _add_key(self, key):
+        keys = list(self._keys)
+        keys.append(key)
+        self._fail_if_duplicate_keys(keys)
```

Comment:
These two lines can become:

```python
self.keys = keys
```

This will perform additional logic missing in the current implementation, e.g. keys need to be cast to an immutable numpy array of object dtype.